03-Binary-Search-Trees/03-Binary-Search-Trees-Height-Big-O.md

```diff
@@ -0,0 +1,357 @@
+
+### !callout-warning
+
+## Tree Balance & Time Complexity
+
+Assume that we are using the Binary Search tree methods as we have described them here.  If all the nodes were inserted in order, the worst case time complexity would be O(n) because each node except the root would be the left child of the previous nodes if the nodes were inserted in descending order, or the right child of the previous node if they were inserted in ascending order.  The tree would be **unbalanced**.  There are more advanced algorithms that can be used to create self-balancing trees.
+
+* If a tree is unbalanced it's time complexities for Insertion, Deletion, and searching approach O(n).
+* If a tree is balanced it's time complexities for Insertion, Deletion, and searching are O(log n).
+
+Therefore it is very important that a tree **remain balanced**.
+
+### !end-callout
+
+### !challenge
+
+* type: number
+* id: 476cab10-a81f-4877-aad3-682602a3c30d
+* title: How many nodes do you need to examine to find 5, in the 1st example?
+* decimal: 0
+* points: 1
+* topics: bst
+
+##### !question
+
+How many nodes do you need to examine to find 5, in the 1st example?
+
+##### !end-question
+
+##### !placeholder
+
+put a number in here
+
+##### !end-placeholder
+
+##### !answer
+
+3
+
+##### !end-answer
+
+
+### !end-challenge
+
+<!-- ======================= END CHALLENGE ======================= -->
+
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+
+### !challenge
+
+* type: number
+* id: 5a4ce281-7f17-47e5-961e-a582171eb3be
+* title: How many nodes to check in the 2nd example?
+* decimal: 0
+* points: 1
+* topics: bst
+
+##### !question
+
+How many in the 2nd?
+
+##### !end-question
+
+##### !placeholder
+
+Number goes here
+
+##### !end-placeholder
+
+##### !answer
+
+5
+
+##### !end-answer
+
+
+### !end-challenge
+
+<!-- ======================= END CHALLENGE ======================= -->
+
+
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+
+### !challenge
+
+* type: number
+* id: b7d69d24-04e0-4891-beaa-ccbb81f73fed
+* title: 5 level tree
+* decimal: 0
+* points: 1
+* topics: bst
+
+##### !question
+
+With the [Binary Tree Visualizer](https://visualgo.net/en/bst), build a **balanced** tree with a height of 5 levels.  How many comparisons do you need to make to find a particular leaf node?
+
+##### !end-question
+
+##### !placeholder
+
+Number goes here
+
+##### !end-placeholder
+
+##### !answer
+
+5
+
+##### !end-answer
+
+### !end-challenge
+
+<!-- ======================= END CHALLENGE ======================= -->
+
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+
+### !challenge
+
+* type: number
+* id: 0dd499dc-7260-4d51-aabd-b9f4560e1047
+* title: 6 level tree
+* decimal: 0
+* points: 1
+* topics: bst
+
+##### !question
+
+Add 5 more nodes to the balanced tree, maintaining the balance.  How many comparisons do you need to make now to find a particular leaf node?
```

Comment:
Honestly I kinda guessed at this one haha. I'm not sure if it can be worded differently, but I found it a little confusing.